# Data Preparation

## 1. Loading the Data

The datasets are loaded again in this notebook because variables created in
`01_schema_validation.ipynb` are not available here.

The original CSV files are loaded using pandas. The validated datasets are
then used to create standardized CSV files for MySQL preparation.

In [2]:
import pandas as pd

customers_df = pd.read_csv("../data/olist_customers_dataset.csv")
orders_df = pd.read_csv("../data/olist_orders_dataset.csv")
order_items_df = pd.read_csv("../data/olist_order_items_dataset.csv")
order_payments_df = pd.read_csv("../data/olist_order_payments_dataset.csv")
order_reviews_df = pd.read_csv("../data/olist_order_reviews_dataset.csv")
products_df = pd.read_csv("../data/olist_products_dataset.csv")
sellers_df = pd.read_csv("../data/olist_sellers_dataset.csv")
geolocation_df = pd.read_csv("../data/olist_geolocation_dataset.csv")
product_category_name_translation_df = pd.read_csv(
    "../data/product_category_name_translation.csv"
)

In [3]:
datasets = {
    "customers": customers_df,
    "orders": orders_df,
    "order_items": order_items_df,
    "order_payments": order_payments_df,
    "order_reviews": order_reviews_df,
    "products": products_df,
    "sellers": sellers_df,
    "geolocation": geolocation_df,
    "product_category_name_translation": product_category_name_translation_df
}

for name, df in datasets.items():
    print(f"{name}: {df.shape}")

customers: (99441, 5)
orders: (99441, 8)
order_items: (112650, 7)
order_payments: (103886, 5)
order_reviews: (99224, 7)
products: (32951, 9)
sellers: (3095, 4)
geolocation: (1000163, 5)
product_category_name_translation: (71, 2)


## 2.1 Preparing the Customers Data

### Problem

The `customers` DataFrame was correctly parsed and validated in pandas and contains 99,441 customer records.

However, loading the original CSV directly into MySQL caused a CSV parsing issue with quoted fields. As a result, some `customer_id` values were imported incorrectly.

### Solution

Since pandas had already parsed the original CSV correctly, the validated DataFrame was exported again as a new CSV file.

This standardizes the CSV formatting without changing the underlying data.

In [4]:
customers_df.to_csv(
    "../data/cleaned/customers_clean.csv",
    index=False,
    na_rep=""
)

pd.read_csv("../data/cleaned/customers_clean.csv").shape

(99441, 5)

## 2.2 Preparing the Orders Data 

### Problem

The `orders` DataFrame was already validated in pandas and contains 99,441 orders.

However, loading the original CSV directly into MySQL caused a CSV parsing issue with quoted fields. As a result, some `order_id` values were imported incorrectly.

### Solution

Since pandas had already parsed the original CSV correctly, the validated DataFrame was exported again as a new CSV file.

This standardizes the CSV formatting without changing the underlying data.

The cleaned file was created using:

In [5]:
orders_df.to_csv(
    "../data/cleaned/orders_clean.csv",
    index=False,
    na_rep=""
)

pd.read_csv("../data/cleaned/orders_clean.csv").shape

(99441, 8)

## 2.3 Preparing the Order Reviews Data

### Problem

The original `order_reviews` CSV contains review text fields that may include:

- commas
- quotation marks
- missing values
- line breaks

Although pandas was able to read the original CSV correctly, loading the same file directly into MySQL produced an incorrect number of rows.

For example, loading the original file with a simple comma separator resulted in more rows than expected because commas inside review text were interpreted as column separators.

Using `ENCLOSED BY '"'` on the original file also did not produce the expected result because the original CSV formatting was not fully compatible with the MySQL `LOAD DATA` settings.

The pandas DataFrame was therefore used as the trusted representation of the data.

### Solution

The validated `order_reviews_df` DataFrame was exported again as a new CSV file.

This preserves the correctly parsed review data while creating a consistently formatted CSV file for MySQL import.

In [7]:
order_reviews_df.to_csv(
    "../data/cleaned/order_reviews_clean.csv",
    index=False,
    na_rep=""
)

pd.read_csv("../data/cleaned/order_reviews_clean.csv").shape

(99224, 7)

## 2.4 Preparing the Order Items Data

### Problem

The original `order_items` CSV contains inconsistent field quoting.
Although pandas correctly parses the file and the ID columns contain no
leading quotation marks, the original CSV formatting may cause parsing
issues during MySQL import.

### Solution

The validated `order_items_df` DataFrame is exported again as a new CSV file.

This standardizes the CSV formatting without changing the underlying data
and provides a consistent file for MySQL import.

In [10]:
order_items_df.to_csv(
    "../data/cleaned/order_items_clean.csv",
    index=False,
    na_rep=""
)

pd.read_csv("../data/cleaned/order_items_clean.csv").shape

(112650, 7)

## 2.5 Preparing the Order Payments Data

### Problem

The original `order_payments` CSV contains inconsistent field quoting.
Although pandas correctly parses the file, the original CSV formatting may
cause parsing issues during MySQL import.

### Solution

The validated `order_payments_df` DataFrame is exported again as a new CSV file.

This standardizes the CSV formatting without changing the underlying data
and provides a consistent file for MySQL import.

In [11]:
order_payments_df.to_csv(
    "../data/cleaned/order_payments_clean.csv",
    index=False,
    na_rep=""
)

pd.read_csv("../data/cleaned/order_payments_clean.csv").shape

(103886, 5)

## 2.6 Preparing the Products Data

### Problem

The original `products` CSV contains inconsistent field quoting.
Although pandas correctly parses the file, the original CSV formatting may
cause parsing issues during MySQL import.

### Solution

The validated `products_df` DataFrame is exported again as a new CSV file.

This standardizes the CSV formatting without changing the underlying data
and provides a consistent file for MySQL import.

In [12]:
products_df.to_csv(
    "../data/cleaned/products_clean.csv",
    index=False,
    na_rep=""
)

pd.read_csv("../data/cleaned/products_clean.csv").shape

(32951, 9)

## 2.7 Preparing the Sellers Data

### Problem

The original `sellers` CSV contains inconsistent field quoting.
Although pandas correctly parses the file, the original CSV formatting may
cause parsing issues during MySQL import.

### Solution

The validated `sellers_df` DataFrame is exported again as a new CSV file.

This standardizes the CSV formatting without changing the underlying data
and provides a consistent file for MySQL import.

In [13]:
sellers_df.to_csv(
    "../data/cleaned/sellers_clean.csv",
    index=False,
    na_rep=""
)

pd.read_csv("../data/cleaned/sellers_clean.csv").shape

(3095, 4)

## 2.8 Validating the Prepared CSV Files

After creating the prepared CSV files, their shapes were compared with the
original validated DataFrames to ensure that no rows or columns were lost
during export.

The prepared files should have the same shape as their corresponding
validated DataFrames.

In [14]:
cleaned_files = {
    "customers": "../data/cleaned/customers_clean.csv",
    "orders": "../data/cleaned/orders_clean.csv",
    "order_items": "../data/cleaned/order_items_clean.csv",
    "order_payments": "../data/cleaned/order_payments_clean.csv",
    "order_reviews": "../data/cleaned/order_reviews_clean.csv",
    "products": "../data/cleaned/products_clean.csv",
    "sellers": "../data/cleaned/sellers_clean.csv"
}

for name, path in cleaned_files.items():
    cleaned_df = pd.read_csv(path)
    print(f"{name}: original={datasets[name].shape}, cleaned={cleaned_df.shape}")

customers: original=(99441, 5), cleaned=(99441, 5)
orders: original=(99441, 8), cleaned=(99441, 8)
order_items: original=(112650, 7), cleaned=(112650, 7)
order_payments: original=(103886, 5), cleaned=(103886, 5)
order_reviews: original=(99224, 7), cleaned=(99224, 7)
products: original=(32951, 9), cleaned=(32951, 9)
sellers: original=(3095, 4), cleaned=(3095, 4)


## 2.9 Preparation Summary

The validated pandas DataFrames were used to create standardized CSV files
for tables that required preparation before MySQL import.

The prepared files are:

- `customers_clean.csv`
- `orders_clean.csv`
- `order_items_clean.csv`
- `order_payments_clean.csv`
- `order_reviews_clean.csv`
- `products_clean.csv`
- `sellers_clean.csv`

All prepared files were saved in the `data/cleaned/` directory.

The row counts and column counts of the prepared files were verified against
the original validated DataFrames.

No data was intentionally removed or modified during this preparation step.